### PR TITLE
bpf: Do not pre-allocate BPF maps by default

### PR DESCRIPTION
**Pull request by tgraf**
Set the daemon option to by default have the same setting as previous
versions of Cilium (enable preallocation where supported).
    
Update the kubernetes YAMLs such that users who upgrade just the
DaemonSet will have minimal upgrade impact as the current maps will be
retained with pre-allocated values. At the same time, add a default
option to the ConfigMap such that new users will automatically deploy
without pre-allocation, minimizing memory usage in newer deployments.
    
For more information on the new setting, see the ConfigMap comments.

Individual commit logs have further information.

Related: #6351 

This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6357)

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

File: bpf/bpf_lxc.c [hunk 1]
```diff
@@ -62,6 +62,7 @@ struct bpf_elf_map __section_maps CT_MAP_TCP6 = {
 	.size_value	= sizeof(struct ct_entry),
 	.pinning	= PIN_GLOBAL_NS,
 	.max_elem	= CT_MAP_SIZE_TCP,
+	.flags		= DEFAULT_MAP_FLAGS,
```
Comment (borkmann) [inline bpf/bpf_lxc.c]: LRU maps in non-prealloc mode cannot be created in the kernel.

Comment (tgraf) [reply bpf/bpf_lxc.c]: Right, I was testing on 4.9 without LRU to get started. I've now made it conditional.

**Comment by tgraf** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

File: bpf/lib/maps.h [hunk 2]
```diff
@@ -173,7 +179,7 @@ struct bpf_elf_map __section_maps cilium_ipcache = {
 	.size_value	= sizeof(struct remote_endpoint_info),
 	.pinning	= PIN_GLOBAL_NS,
 	.max_elem	= IPCACHE_MAP_SIZE,
-	.flags		= BPF_F_NO_PREALLOC,
+	.flags		= CONDITIONAL_PREALLOC,
```
Comment (borkmann) [inline bpf/lib/maps.h]: This has to be fixed BPF_F_NO_PREALLOC as LPM map doesn't support prealloc.

**Review by borkmann — APPROVED** [review]
Aside from the one issue lgtm

**Comment by joestringer** [comment]
The approach where the option flags are checked in the `bpf.NewMap(...)` invocations doesn't really work, because this tends to run on application init rather than after the daemon's commandlines have been parsed. I'll shift that logic into `OpenOrCreateMap(...)`.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Realizing that the trick I did above is still not enough, the `OpenAfterMount()` is still being called from golang init which leads to reading the wrong value out from the setting before the setting is updated as per the commandline arguments. These calls will all need to be blocked by the main thread until after commandline parsing.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
There's still something strange occurring, at least with lxcmap, where it ends up empty (implying that even the controller isn't pushing into the map that's actually pinned to the filesystem). I don't think that the new "preallocation disabled" mode works properly for hashtables right now, but I don't yet know why.

**Comment by joestringer** [comment]
Found a mismatch in the policymap creation flags due to policymap using its own custom code for map management, Fixed that issue.

Next issue is that when toggling this flag on or off (edit configmap + `kubectl -n kube-system delete po -l k8s-app=cilium`), endpoints are not restored from previous run. Pods must be restarted to reattach to Cilium. Not sure why yet.

EDIT: This appears to be something peculiar to my environment---My kubelet uses different docker settings than what Cilium would by default. Cilium only invokes its own docker client when restoring endpoints so I didn't notice that there's a mismatch before.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

EDIT: Green runs from k8s-1.13 and runtime, k8s-1.8 lost connectivity to the node:
```
22:19:36 Cannot contact jenkins-fixed-3: java.lang.InterruptedException
```

We also hit an issue during downgrade to v1.3 where etcd could not be contacted for an extended period of time. Flake?

https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/8616/testReport/junit/k8s-1/8/K8sUpdates_Tests_upgrade_and_downgrade_from_a_Cilium_stable_image_to_master/

```
2018-12-19T22:48:33.218512896Z level=fatal msg="timed out while waiting for etcd session. Ensure that etcd is running on [https://cilium-etcd-client.kube-system.svc:2379]" subsys=kvstore
```

**Comment by joestringer** [comment]
test-me-please

EDIT: Same upgrade test failure. Summary is that v1.3.0 is deployed, then upgraded to this PR (which appears to be OK), then downgraded again to v1.3.0, at which point cilium loses access to etcd.

From `cilium-qfpjr` status:
```
KVStore:                Failure   Err: Not able to connect to any etcd endpoints - etcd: 0/1 connected: https://cilium-etcd-client.kube-system.svc:2379 - context deadline exceeded
```

From one of the kube-dns pod logs:
```
2018-12-20T00:21:01.535558954Z I1220 00:21:01.535234       1 dns.go:670] No record found for cilium-etcd-client.kube-system.svc.cluster.local.
2018-12-20T00:21:01.535562164Z I1220 00:21:01.535237       1 dns.go:670] No record found for cilium-etcd-client.kube-system.svc.cluster.local.
```

I also note that kube-dns has restarted 7 times.

https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/8624/testReport/k8s-1/8/K8sUpdates_Tests_upgrade_and_downgrade_from_a_Cilium_stable_image_to_master/

[0c9fd4ca_K8sUpdates_Tests_upgrade_and_downgrade_from_a_Cilium_stable_image_to_master.zip](https://github.com/cilium/cilium/files/2700666/0c9fd4ca_K8sUpdates_Tests_upgrade_and_downgrade_from_a_Cilium_stable_image_to_master.zip)

**Comment by joestringer** [comment]
In local testing, I'm yet to reproduce this but am hitting all sorts of other weird and interesting behaviours. Propagation of service DNS names into kube-dns often takes multiple minutes, and I've seen the cilium endpoints ready wait command (`kubectl.CiliumEndpointWaitReady()`) do nothing for about 30 minutes, even though when logging into each cilium pod, every endpoint is ready. Currently considering changing the version we handle upgrade/downgrade to v1.3.1 in case it improves any of the behaivours. Hard to really say whether we have a legitimate downgrade breakage or not just yet.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Build timed out:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/8818/execution/node/97/log/

```
23:19:15 K8sUpdates 
23:19:15   Tests upgrade and downgrade from a Cilium stable image to master
23:19:15   /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:383
23:21:04 STEP: Installing a cleaning state of Cilium
23:21:04 STEP: Installing kube-dns
23:21:04 STEP: Deploying etcd-operator
23:21:04 STEP: Cilium "v1.3" is installed and running
23:21:15 STEP: Performing Cilium preflight check
23:21:15 STEP: Checking that no controllers in Cilium have failed
23:21:16 STEP: Checking that cilium-health status is healthy
23:21:21 STEP: Checking that Kubernetes services are correctly plumbed by Cilium
23:21:21 STEP: Checking that the kubernetes service is correctly plumbed by Cilium
23:21:21 STEP: Waiting for all etcd-operator pods are ready
23:22:18 STEP: Installing Microscope
23:22:18 STEP: Creating some endpoints and L7 policy
23:22:19 STEP: Checking that the kubernetes service is correctly plumbed by Cilium
23:22:26 STEP: Waiting for Cilium endpoints to become ready (1)
23:24:45 STEP: Waiting for kube-dns to become ready
23:24:48 STEP: Checking that the kubernetes service is correctly plumbed by Cilium
23:24:48 STEP: Waiting for kube-dns service entries (1)
23:24:48 STEP: Waiting for Cilium endpoints to become ready (2)
23:24:51 STEP: Waiting for kube-dns service entries (2)
23:24:51 STEP: Making L7 requests between endpoints
23:24:51 STEP: Updating cilium to master image
23:24:51 STEP: Install Cilium pre-flight check DaemonSet
23:24:51 STEP: Waiting for all cilium pre-flight pods to be ready
23:25:07 STEP: Removing Cilium pre-flight check DaemonSet
23:27:14 STEP: Checking that installed image is "k8s1:5000/cilium/cilium-dev:latest"
23:27:14 STEP: Waiting for Cilium endpoints to become ready (1)
23:31:54 STEP: Waiting for kube-dns to become ready
23:31:54 STEP: Checking that the kubernetes service is correctly plumbed by Cilium
23:31:54 STEP: Waiting for kube-dns service entries (1)
23:31:54 STEP: Waiting for Cilium endpoints to become ready (2)
23:31:54 STEP: Waiting for kube-dns service entries (2)
23:31:54 STEP: Making L7 requests between endpoints
23:31:54 STEP: Downgrading cilium to v1.3 image
23:32:19 STEP: Checking that installed image is "v1.3"
23:32:19 STEP: Waiting for Cilium endpoints to become ready (1)
23:52:21 Cannot contact jenkins-fixed-0: java.lang.InterruptedException
00:07:52 Sending interrupt signal to process
00:07:52 Sending interrupt signal to process
```

It's taking over 2 minutes after Cilium v1.3 + operator resources are created before the cilium endpoints are ready, then over 4.5 minutes with this PR to wait for endpoints to be ready. Removing pre-flight daemonset also notably takes more than 2 minutes. After downgrade from this PR to v1.3 again, we appear to hit some exception before the endpoints become ready, indicating some sort of downgrade issue (endpoints are not ready after 20 minutes).

**Comment by joestringer** [comment]
test-me-please

EDIT: Weird seemingly unrelated error during bootstrap:
```
19:50:27     k8s1-1.13: [bootstraptoken] creating the "cluster-info" ConfigMap in the "kube-public" namespace
19:50:27     k8s1-1.13: error execution phase addon/coredns: unable to create deployment: 0-length response
19:50:27 The SSH command responded with a non-zero exit status. Vagrant
19:50:27 assumes that this means the command failed. The output for this command
19:50:27 should be in the log above. Please read the output to determine what
19:50:27 went wrong.
```

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

EDIT: At this point I updated the configmap in test/k8sT/manifests/ to enable preallocation of BPF maps. Failure appears to be regarding connection disruption during restart:

```
22:01:34 • Failure [326.303 seconds]
22:01:34 K8sChaosTest
22:01:34 /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:351
22:01:34   Restart with long lived connections
22:01:34   /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:351
22:01:34     TCP connection is not dropped when cilium restarts [It]
22:01:34     /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:383
22:01:34 
22:01:34     Failed while cilium was restarting
22:01:34     Expected command: kubectl exec -n default netperf-client -- netperf -l 300 -t TCP_STREAM -H 10.10.0.201 
22:01:34     To succeed, but it failed:
22:01:34     Exitcode: 1 
22:01:34     Stdout:
22:01:34      	 MIGRATED TCP STREAM TEST from 0.0.0.0 (0.0.0.0) port 0 AF_INET to 10.10.0.201 (10.10.0) port 0 AF_INET
22:01:34     	 netperf: send_omni: send_data failed: Connection reset by peer
22:01:34     	 send_data: data send error: Connection reset by peer (errno 104)
22:01:34     	 command terminated with exit code 1
22:01:34     	 
22:01:34     Stderr:
22:01:34      	 
22:01:34     
22:01:34 
22:01:34     /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/k8sT/Chaos.go:237
```

**Comment by joestringer** [comment]
I ran this test locally and it passed, still digging around.

**Comment by joestringer** [comment]
test-me-please

EDIT: Green build.

File: examples/kubernetes/templates/v1/cilium-cm.yaml [hunk 3]
```diff
@@ -70,6 +70,21 @@ data:
   ct-global-max-entries-tcp: "524288"
   ct-global-max-entries-other: "262144"
 
+  # Pre-allocation of map entries allows per-packet latency to be reduced, at
+  # the expense of up-front memory allocation for the entries in the maps. The
+  # default below value will minimize memory usage in the default installation;
```
Comment (tgraf) [inline examples/kubernetes/templates/v1/cilium-cm.yaml]: value below

File: examples/kubernetes/templates/v1/cilium-cm.yaml [hunk 4]
```diff
@@ -70,6 +70,21 @@ data:
...
+  # users who are sensitive to latency may consider enabling this option.
+  #
+  # If this value is modified, then during the next Cilium startup the restore
+  # of existing endpoints and tracking of ongoing connections may be disrupted.
+  # This may lead to policy drops or a change in loadbalancing decisions for a
+  # connection for some time. Endpoints may need to be recreated to restore
+  # connectivity.
+  #
+  # For users upgrading from Cilium 1.3 or earlier, to minimize disruption
```
Comment (tgraf) [inline examples/kubernetes/templates/v1/cilium-cm.yaml]: Let's rephrase this to say that Cilium <= 1.3 had this value to be always true so setting it to false and upgrading from <= 1.3 to >= 1.4 will cause some one-time disruptions during the upgrade. The way it is written right now, people may mistake it as for causing disruptions continuously.

Let's also specify that for versions < 1.4, this option has no effect. We should do this for other fields as well but doesn't hurt to start with this one.

**Comment by joestringer** [comment]
test-me-please

File: examples/kubernetes/templates/v1/cilium-cm.yaml (hunk 4 above)
Comment (joestringer) [reply examples/kubernetes/templates/v1/cilium-cm.yaml]: @tgraf I adjusted for this feedback, PTAL.

File: daemon/daemon.go [hunk 5]
```diff
@@ -556,34 +556,8 @@ func (d *Daemon) init() error {
...
-			if _, err := lbmap.RevNat4Map.OpenOrCreate(); err != nil {
-				return err
-			}
-			if _, err := lbmap.RRSeq4Map.OpenOrCreate(); err != nil {
-				return err
-			}
-			if _, err := proxymap.Proxy4Map.OpenOrCreate(); err != nil {
-				return err
-			}
+		if err := openServiceMaps(); err != nil {
```
Comment (ianvernon) [inline daemon/daemon.go:559]: Thanks for putting this into a separate function!

File: daemon/loadbalancer.go [hunk 6]
```diff
@@ -426,6 +427,40 @@ func (d *Daemon) RevNATDump() ([]loadbalancer.L3n4AddrID, error) {
...
+		if _, err := lbmap.Service4Map.OpenOrCreate(); err != nil {
+			return err
+		}
+		if _, err := lbmap.RevNat4Map.OpenOrCreate(); err != nil {
+			return err
+		}
+		if _, err := lbmap.RRSeq4Map.OpenOrCreate(); err != nil {
+			return err
+		}
+		if _, err := proxymap.Proxy4Map.OpenOrCreate(); err != nil {
```
Comment (ianvernon) [inline daemon/loadbalancer.go:456]: Only slight nit I have here is that proxy maps are unrelated to the loadbalancer code, but I think having this code in a separate function is better than how it is currently in master.

File: examples/kubernetes/templates/v1/cilium-cm.yaml [hunk 7]
```diff
@@ -70,6 +70,24 @@ data:
...
+  # default value below will minimize memory usage in the default installation;
+  # users who are sensitive to latency may consider setting this to "true".
+  #
+  # This option was introduced in Cilium 1.4. Cilium 1.3 and earlier ignore
+  # this option and behave as though it is set to "true".
+  #
+  # If this value is modified, then during the next Cilium startup the restore
+  # of existing endpoints and tracking of ongoing connections may be disrupted.
+  # This may lead to policy drops or a change in loadbalancing decisions for a
+  # connection for some time. Endpoints may need to be recreated to restore
```
Comment (ianvernon) [inline examples/kubernetes/templates/v1/cilium-cm.yaml:84]: This implies that the state is not able to be recovered over time? Or is it just that because the applications may get into a bad state themselves vs. in a bad state on the Cilium side?

**Review by ianvernon — APPROVED** [review]

**Comment by tgraf** [comment]
I can't approve formally as I've contributed commits. So LGTM via comment.

File: examples/kubernetes/templates/v1/cilium-cm.yaml (hunk 7 above)
Comment (joestringer) [reply examples/kubernetes/templates/v1/cilium-cm.yaml:84]: ~~Basically Cilium completely forgets about endpoints that were previously managed, and they must be restarted in order for the new instance of Cilium to start managing them.~~

EDIT: The above is wrong, that was with `clean-cilium-state` enabled in the ConfigMap.

In local upgrade testing of just the Cilium component, it surprisingly seems to update pretty seamlessly (admittedly without testing services, which I'd expect to break due to loadbalancing decisions being lost). I suspect that in the upgrade test, because the etcd instance is being redeployed, some more serious problems occur during bootstrap which take some time to get back into a good state.
